mne/preprocessing/_regress.py

```diff
@@ -53,35 +69,180 @@ def regress_artifact(inst, picks=None, picks_artifact='eog', betas=None,
-        if estimate:
-            # subtract mean over time from every trial/channel
+        # Create an EOGRegression object and load the given betas into it.
+        picks = _picks_to_idx(inst.info, picks, exclude=exclude, none='data')
+        picks_artifact = _picks_to_idx(inst.info, picks_artifact)
+        want_betas_shape = (len(picks), len(picks_artifact))
+        _check_option('betas.shape', betas.shape, (want_betas_shape,))
+        r = EOGRegression(picks, picks_artifact)
+        r.info = pick_info(inst.info, picks)
+        r.coef_ = betas
+        r._picks = r.info['ch_names']
+        r._exclude = exclude
+        r._picks = r.info['ch_names']
+        r._picks_artifact = [inst.ch_names[ch] for ch in picks_artifact]
+        betas = r
+    return betas.apply(inst, taa=taa, copy=copy), betas.coef_
+
+
+@fill_doc
+class EOGRegression():
+    """Remove EOG artifact signals from other channels by regression.
+
+    Employs linear regression to remove signals captured by some channels,
+    typically EOG, but it also works with ECG from other channels, as described
+    in :footcite:`GrattonEtAl1983`. You can also chose to fit the regression
+    coefficients on evoked blink/saccade data and then apply them to continuous
+    data, as described in :footcite:`CroftBarry2000`.
+
+    Parameters
+    ----------
+    %(picks_good_data)s
+    exclude : list | 'bads'
+        List of channels to exclude from the regression, only used when picking
+        based on types (e.g., exclude="bads" when picks="meg").
+        Specify ``'bads'`` (the default) to exclude all channels marked as bad.
+    picks_artifact : array-like | str
+        Channel picks to use as predictor/explanatory variables capturing
+        the artifact of interest (default is "eog").
+
+    Notes
+    -----
+    .. versionadded:: 1.2
+
+    References
+    ----------
+    .. footbibliography::
+    """
+
+    def __init__(self, picks=None, exclude='bads', picks_artifact='eog'):
+        self._picks = picks
+        self._exclude = exclude
+        self._picks_artifact = picks_artifact
+
+    def fit(self, inst):
+        """Fit EOG regression coefficients.
+
+        Parameters
+        ----------
+        inst : Raw | Epochs | Evoked
+            The data on which the EOG regression weights should be fitted.
+
+        Returns
+        -------
+        self : EOGRegression
+            The fitted ``EOGRegression`` object. The regression coefficients
+            are available as the ``.coef_`` and ``.intercep_`` attributes.
+
+        Notes
+        -----
+        If your data contains EEG channels, make sure to apply the desired
+        reference (see :func:`mne.set_eeg_reference`) before performing EOG
+        regression.
+        """
+        self._check_inst(inst)
+        picks = _picks_to_idx(inst.info, self._picks, none='data',
+                              exclude=self._exclude)
+        picks_artifact = _picks_to_idx(inst.info, self._picks_artifact)
```

Comment:
Yes, because the `idx` could be different between `fit` and `apply`. To provide some form of flexibility, channel names are used to match channels, not integer indices. Common use case: fitting to evoked that have no stim channel, applying to raw which has.

Reply:
ah, that makes sense. clever.